### ISSUE TITLE
LuaSkin vs idiomatic Lua 😀 

### ISSUE THREAD

**Issue by lowne** [Hammerspoon/hammerspoon#690]
`[18:31:22]  <lowne_>	LuaSkin's toNSObjectAtIndex is clearly to blame 😊`

**Comment by lowne** [comment]
Since `hs.settings.clear(key)` seems to work, amending the docs would be a valid "fix" for the error, but not for the hanging.

**Comment by asmagill** [comment]
Huh?

~~~lua
> hs.settings.set("me","you")
-- Loading extension: settings


> hs.settings.get("me")
you

> hs.settings.set("me",nil)


> hs.settings.get("me")
nil
~~~

**Comment by lowne** [comment]
I didn't have a particularly outdated `master`, but at any rate with the current master the hang is gone; however:
```
> settings.set('me',nil)

> settings.get('me')
nil

> settings.set('me')

> settings.get('me')
no value: 0x0
```

**Comment by lowne** [comment]
OT: @asmagill  I suspect I could have messed up the build somehow; I'm now using your build script, and I was wondering if the line
```
rm -fr `xcodebuild -workspace Hammerspoon.xcworkspace -scheme Hammerspoon -configuration DEBUG -showBuildSettings | sort | uniq | grep " BUILT_PRODUCTS_DIR =" | awk '{ print $3 }'`/Hammerspoon.app
```
is really supposed to go *after* `make`?

**Comment by asmagill** [comment]
The changed title makes it more clear... I missed that you were leaving off a value for set... settings pre-dates a lot of LuaSkin, so it doesn't do an argument check at the beginning like almost all newer modules, it just attempts to use the value at argument position 2 as the value to assign to the key name given... accessing a missing variable (i.e. not provided) with the C_API is equivalent to "nil" in *most* cases, but not all... I wonder if, depending upon what's been on the lua stack recently, it's picking up garbage from one of the edge cases...

The proper fix, of course, is to either require 2 args for `set` (i.e. an explicit nil would be expected) or to note the absence and internally treat it as nil...  which behavior is preferred?

And as for the rm line, yes, if you're using the build script I use and don't have a modified make file, the make file copies the built Hammerspoon into the source build/ directory... this line removes the app from the derived_data directory XCode used during the actual build.  I added this because depending upon how I started Hammerspoon, it was sometimes starting the "wrong" one, screwing up paths.

**Comment by asmagill** [comment]
Thinking about it more, that has to be whats happening... toNSObjectAtIndex converts the index passed to an absolute index so it can be used for more than just arguments being passed to a function/method... since its an absolute index, lua's internal protections around psuedo indexes aren't in place and there is no way for LuaSkin to realize it's just been passed a valid but unacceptable index.

So, yeah, which behavior should be enforced: require 2 args, or explicit nil when given only one arg?

**Comment by lowne** [comment]
I'd say the latter, given that requiring a nil as the *last* argument in a function call really isn't idiomatic Lua.

**Comment by asmagill** [comment]
Just to clarify, you want to change the behavior of `hs.settings.set`, correct?

According to the docs:

~~~
/// hs.settings.set(key, val)
/// Function
/// Saves a setting with common datatypes
///
/// Parameters:
///  * key - A string containing the name of the setting
///  * val - A value for the setting. Valid datatypes are:
///   * string
///   * number
///   * boolean
///   * nil
///   * table (which may contain any of the same valid datatypes)
///
/// Returns:
///  * None
///
/// Notes:
///  * This function cannot set dates or raw data types, see `hs.settings.setDate()` and `hs.settings.setData()`
~~~

using `set` without a second parameter should be an error.  I can go either way, but I want to be clear about what change we're making.

**Comment by lowne** [comment]
Yes.
If the docs don't explicitly mention the error, then it's not expected; and if they did, it'd still be non-idiomatic, as `nil` is the *absence* of a value (as opposed to e.g. None in python); but even thinking about it as an "existing" value, given that:
```lua
a,b,c = 1 -- b and c "are" nil
function f(a,b,c) return 1 end
f(1) -- f's b and c "are" nil
a,b,c=f() -- b and c "are" nil
```
having `f(a,nil)~=f(a)` is, at the very least, extremely surprising.

**Comment by lowne** [comment]
Generally speaking, if that's not already the case, LuaSkin's `getArgumentAtIndex(i)` functions should always return nil when `i>#args`

**Comment by asmagill** [comment]
That's something I never fully understood about lua... its so easy to distinguish between nil and none on the C side, but all but impossible to do so on the lua side...

Whatever, the patch supports your usage.

**Comment by lowne** [comment]
`settings.set` is a particularly prominent case, as I (and I assume others) use it quite often from the console (omitting the 2nd argument, as one does) for testing etc. This issue is less of a problem for scripts, as calling `settings.set(a,b)` where b can be nil still results in two arguments being passed. (Note that I'd still use `settings.set(a)` if I just wanted to clear `a` somewhere in my config). However, as I said, the fix should go (if possible) in LuaSkin itself, as presumably there might be similar cases elsewhere.

**Comment by asmagill** [comment]
It can't go in LuaSkin.  The nature of the LuaSkin methods we're talking about is that they can be useful in C functions (those we advertise to Hammerspoon/Lua) and in callbacks/delegates/etc (the C-side, before/after the Hammerspoon function is called), which do not have the protection of the Lua API to isolate their "virtual stack" from the rest of the Lua state.  It is this very protection that allows lua to be "lazy" and "assume none and nil are equivalent", both in the Lua language proper, and the C additions from most modules.

What perhaps does need to be done, however, is better documentation (so far, its mostly what @cmsj and I have put into the LuaSkin.h file).

For what it's worth, were I to code settings today, the set function would have started with using LuaSkin to check it's arguments before doing anything with the actual stack, and it would return an error if less than 2 arguments were provided... but the check wasn't available when this was first written, plus the original didn't use [[LuaSkin shared] toNSObjectAtIndex:#] either (it used a whole lot of ugly, redundant, and unscalable code), so the "virtual stack" enforced by the lua API protected things if you happened to leave argument 2 out.  Its a trade off, but one that needs to be more clearly documented for certain cases.

The submitted pull makes the second argument optional, and explicitly defaults to nil if it's absent, so your use case stands.

**Comment by lowne** [comment]
I accept and appreciate that, and consider the issue solved; what follows is for pure intellectual curiosity:
re: forcing `settings.set` to take two arguments, then ofc you'd need to disallow nil and enforce use of `settings.clear` for that, else we'd be back to `f(a,nil)~=f(a)`. It'd still feel non-idiomatic though (to a lesser degree), because it'd be akin to require `table.clear(tbl,'field')` instead of `tbl.field=nil`.

About LuaSkin:
>  LuaSkin's getArgumentAtIndex(i) functions should always return nil when i>#args

Why can't that translate, at the top of `toNSObjectAtIndex(idx)` (etc.), to `if (lua_gettop(_L) < idx) return nil;`? (well, in this case, it should return `[NSNull null]` I assume; a hypothetical toBoolean would return false, etc.). From what you wrote, it seems that `lua_gettop` wouldn't work outside of a "virtual stack", but I'm not sure I understand why the stack is different for LuaSkin; if that's related to the `lua_absindex` conversion, the `lua_gettop` check would simply go before that, no?

**Comment by lowne** [comment]
FWIW I tried just that, and:
```lua
> settings.get'me'
no value: 0x0

> settings.set'me'


> settings.get'me'
nil
```

**Comment by lowne** [comment]
To clarify what I did:
```
diff --git a/LuaSkin/LuaSkin/Skin.m b/LuaSkin/LuaSkin/Skin.m
@@ -720,4 +720,5 @@ nextarg:
        allowSelfReference:(BOOL)allow {

+    if (lua_gettop(_L) < idx) return [NSNull null];
     int realIndex = lua_absindex(_L, idx) ;
     NSMutableArray *seenObject = [alreadySeen objectForKey:[NSValue valueWithPointer:lua_topointer(_L, idx)]] ;
```

**Comment by asmagill** [comment]
nil *can* be a valid value for a parameter... just because lua makes it easy for f(1,nil) == f(1) doesn't mean that it *has* to.  And on the C side, none and nil are 2 separate types and very easy to differentiate. It's possible to do so in pure Lua (with a lot of ... and table.pack()), but annoyingly tedious, so most people don't, hence the convention and assumption by many that a trailing nil is *always* optional and that that references to arguments not present should *always* be nil or extra arguments should *always* be ignored.

As to your latest example code, posted while I was responding to this, the simplest example where this assumption fails is when idx is passed in as a negative number.  In that case, the top will *always* be greater.  You *might* be able to get away with lua_absindex(L, lua_gettop(L)) < lua_absindex(L, idx), but I'd have to run tests... It still feels iffy to me.  You're assuming we're looking at arguments being passed in, not an element of a table in a table in a reference stored in the registry... try guessing what that index number might be and when it's valid or invalid...

Because luaskin has no assurance that it is called from within the protected confines of a self cleaning "virtual stack" provided by lua_call/lua_pcall (i.e. the lua interpreter), it's still a hell of a lot safer to do bounds checking in the caller and let luaskin assume that if it's passed an index, it should use that index... it can't know the state of how/where it was called and can't *know* that you want a nil, unless it sees a nil.

And since were discussing it, honestly, I'd prefer that it never assume it knows I mean nil when something doesn't make sense/is out of bounds... it's a whole lot easier to know that I have a bug to track down when I get a "can't index function" error rather than a nil... which might be a valid answer, or might be an assumption.

**Comment by lowne** [comment]
Well, this conversation is getting more and more abstract, but I'm one of those obnoxious people that genuinely enjoy abstractness, so here it goes :)

First: the whole point of Lua is simplicity, and Lua's `nil` is certainly among the top reasons Lua is *simple*; it's a bit like memory allocation: to a spreadsheet user, an empty cell is an empty cell; of course there's no such thing as an empty *RAM* cell, therefore we have 50 layers in-between that jump through endless hoops to keep track of things on the side and give the user the "illusion" of an empty cell. Which is also what the Lua *implementation* does (alas not perfectly, see below). This doesn't mean that *nobody* could ever prefer a different, non-"nanny"-spreadsheet that doesn't "lie" and would show zeroes (or, more likely, random garbage) instead of emptiness; for example many embedded programmers, I gather, genuinely love that sort of stuff, and are well compensated for their mastery of PEEK and POKE :D

(As an aside (talking of Lua-the-language here): afaict the *only* reason it's ever possible to tell apart "implied" and "explicit" nils is to support explicit varargs notation (which itself could well *not* be part of the language, since one can always pass tables); and therefore it's only possible to do so within varargs functions (the debug module, as the name implies, shouldn't be considered part of the syntax). In short: it's certainly a case of less-than-ideal leaky abstraction; "a trailing nil being always optional" is not a mere convention - it's the *way the language works*. You *could* decide to write Lua with *only* varargs functions and *never* using multiple assignment, but then you'd be just using the wrong language.)

Second: I *assume* that the point of HS and its predecessor(s) is to make it *simple* to script/extend/customize its functionality; therefore, Lua. (If the point was, say, pure performance, then the way to "script" it would be to fork, code (in C), compile; if the point was to build large scale complex architectures on top of it, the scripting should rather be done in Python; etc.). And once it's decided the *user* API is meant to be used in ("pure") Lua, I (again) *assume* that ideally the API itself should conform to the language; any divergence must be considered another leaky abstraction; and I further assume that leaky abstractions are bad, if sometimes unavoidable - but if they *can* be avoided, they should. 

There have been several instances already where I (so to speak) complained about leaky abstractions or advocated for more "Lua-conformant" solutions; since I'm not the one dealing with Obj-C, I absolutely accept both "can't be done" and "it's a bit of a PITA to implement it that way, so let's not do it", but I've been surprised by arguments that boil down to "the way it's done in C/Obj-C/Cocoa feels better/more natural, therefore that's what the user API should look like" more than once. But it's still a perfectly *valid* argument: it just means that my assumptions in the above paragraph are wrong; in that case I'll simply point out that 

1. these divergencies, lacking further information, are *unexpected* from the user's perspective (who knows nothing about the underlying implementation choices), and should therefore be made explicit and highlighted in the relevant documentation

2. the end product might be easier to work with (for those that actually write the implementation code), but it's certainly less *consistent* for end users; we should be aware (and accepting) of this tradeoff

(end abstract pseudorant)

-----

Lastly:
> You're assuming we're looking at arguments being passed in

Yes, I thought `toNSObjectAtIndex` would *only* be used to get function arguments, but of course I'm not remotely sufficiently familiar with the C codebase to know that for sure. My (fairly large, complex, and full of callbacks) config is working just fine, but I agree that means nothing. Still: if that's not the case, it doesn't seem inconceivable to have separate argument-retrieving-only methods that do the nil trick, and reference-retrieving methods that don't.

> I'd prefer that it never assume it knows I mean nil when something doesn't make sense/is out of bounds

I can relate to that (I've done my share of assembly/C++/Java/Python); but Lua requires - I'd say *enables* - a different - I'd say *simpler* - kind of thinking (see rant above)

**Comment by asmagill** [comment]
you and I apparently don't read the same comments on the lua blogs/forums... I can't count high enough to reach the number of times I've seen someone complaining about the fact that things are made harder to debug because Lua doesn't complain about undefined variables and assumes nil where something is missing... it makes debugging exponentially more time consuming and difficult to do because an obvious error at point "a" could prevent all kinds of "almost" behavior that causes an error ultimately at point "g".

Most modules being added now do explicit type checking at the beginning of each function, a feature added to LuaSkin as well... I didn't intend to spark a long conversation by my comment about how I'd code hs.settings.set today, other than to note that today, I would have started the function with `[[LuaSkin shared] checkArgs:LS_TSTRING, LS_TANY, LS_TBREAK]` where as to maintain the behavior you wished, I had to do `[[LuaSkin shared] checkArgs:LS_TSTRING, LS_TANY | LS_TOPTIONAL, LS_TBREAK]` and then include a check to see if that 2nd arg was present or not.  It's not a hard, just different set of assumptions that flavor the rest of the function.  And since I knew/know that a `hs.settings.clear` does exist... well the idea that set would ever take less than 2 args would seem foreign to me.

And if you want to bring intuitive into it... `hs.settings.set("var")`.  Set var to what? "On"? (which in computerese usually means "True" or 1).  Then where is the unset? See, your assumption isn't intuitive to me :-) Better to be explicit except where its safe to be vague... and anything whose ultimate destination is *not* the Lua interpreter, better not be vague, at least in my opinion and experience.

**Comment by lowne** [comment]
Let's talk intuitive :) - imagine this:

```lua
> a=5
a=5
> b=a+3
b=8
> a=10
a=10
> b
b=13
```
Me, you, or any programmer would probably say "wtf?" at that last line, or, at best, "this must be a reactive implementation that jumped the shark"; whereas, to bring back the basic-level spreadsheet user, they'd say "duh, that's Excel, you're just writing down everything you do for some reason" (fwiw that bit above is one `setmetatable` away from being actual valid Lua, with a "reactive" module I made not long ago).

-----
>  assumes nil where something is missing [...] `hs.settings.set("var")`. Set var to what? "On"?

Again: nil **is** the absence of value. This notion is part of the "different" kind of thinking I mentioned at the end of my previous post. If you reject it (which seems indicated by thinking of nil-for-missing-values as "assumed") for whatever reason (and a *very* good reason to reject it would be that you spend a significant amount of time working with more sensible languages), then of course it's the opposite of intuitive; it seems like a capricious, arbitrary rule. And conversely: set var to what? Once you "know" that nil is the absence of value, the question itself is utterly redundant: to nil, obviously - there's nothing there!

> someone complaining [...] because Lua doesn't complain about undefined variables

Oh, I've seen that; but the problem is completely orthogonal to the "nature of nil", as evinced by the fact that the solutions for it are quite [trivial](https://google.com/search?q=lua+strict+mode) *within* Lua. Moreover, countless people (me included) also complain that *all* dynamic languages make debugging so much harder because of lack of type safety; and many others complain of unpredictable performance for their inner loops in liberally-garbage-collected languages/[platforms](https://android.com); and some more complain about so-called compiler optimisations and doggedly write their algorithms in assembly (with sometimes [interesting](https://randomascii.wordpress.com/2012/12/29/the-surprising-subtleties-of-zeroing-a-register/) [results](https://en.wikipedia.org/wiki/Michael_Abrash#Technical_writer)); and a case has ([successfully](https://arm.com)) been made that the x86 microarchitecture itself is too "abstract".
To which I say ([to](https://github.com/andremm/typedlua) [myself](https://www.haskell.org) [as](https://coq.inria.fr) [well](https://en.wikipedia.org/wiki/Lambda_cube)) that if you're *really* convinced that you're using a given language *despite* as opposed to *because*, it *must* be the case that you have no other choice. Otherwise, just pick a better language :)

_____
Finally, for further appreciation of the difference in perspective, this is how I would write `hs.settings` - what's the most Lua-like API? Just a plain old table that happens to be persisted behind the scenes:

```lua
-- c-like (no optionals down here, please) settings.get,.set,.clear:
local getProp,setProp,clearProp = 
  hs._getProperty,hs._setProperty,hs._clearProperty 
  
return setmetatable({},{
  __index=function(t,k)
    return getProp(k)
  end,
  __newindex=function(t,k,v)
    if v==nil then return clearProp(k)
    else return setProp(k,v) end
  end
})
```

Would you then still object to `hs.settings.var=nil` as unintuitive? What am I setting it to?

**Comment by asmagill** [comment]
TL;DR version: You're wrong, I'm right :-)  Seriously, though, I do agree that perspective can color one's stance, so here are some of my thoughts on why I have my stance...

- - -

As you say, this has gotten far afield, and I will agree to assumptions making or breaking a given interpretation, so... we're not going to come to a full agreement :-), but

> what's the most Lua-like API? Just a plain old table that happens to be persisted behind the scenes:

That's a wrong assumption to make (imo)... you're not dealing with lua when you deal with hs.settings (or most, but definitely not all, of the Hammerspoon modules).  You're dealing with a bridge to OS X (specifically to the User Defaults library.  FWIW, you can actually do this:
~~~ lua
> hs.settings.get("AppleInterfaceStyle")
Dark
~~~
to tell if dark mode is enabled... that's nowhere in Hammerspoon or Lua... that's in NSGlobalDomain, which is inherited by *every* application -- Since we're not sandboxed, we can even inherit other user default databases, like `com.apple.Spaces` (I found that it sometimes lagged a little behind what I was able to get with the updated `undocumented.spaces` module, so I never really went anywhere with it)  But I don't see a real need (yet, perhaps) to extend settings to allow selecting an arbitrary database to add to the search path.)

There will always be some loss or give-and-take.  Why do you think I banged my head against the Objective-C interpretation of UTF8 (as whole, single entity characters) versus the Lua interpretation as "a string of 1 to 4 bytes"?  This makes anything which returns or accepts an index position for a "string like object" to/from lua a real pain in the butt to code... once I'm a little more comfortable that I haven't missed something in there, I may add that to LuaSkin as well so no one else has to bang their head against it, but for the moment, its mostly affected things I've added, so I've felt no rush.  There is code in hs.styledtext, hs.console, MJLua.m (the basic builtin functions that are part of the `hs`), and even hs.speech specifically to deal with this difference, as well as the limitation of the Console to not be able to display anything that isn't a "valid" UTF8 sequence without jumping through hoops (if you use the `hs` command line tool, you may notice that it's output and the console's output differ if you're dealing with non-UTF8 binary data in strings since, like the lua core, it also doesn't suffer from that limitation).

We can certainly disagree over where the loss/give-and-take should go and how much we should do to hide/simplify it, but... it's an integral fact that Hammerspoon is not and never will be *just* Lua.  It's a hybrid of Lua and a bridge to some of OS X.

We can't hide that Hammerspoon isn't just Lua when you're dealing with the bridges to OS X components; at best we can decide which ones to simplify; I prefer giving access to as much as possible, but then wrap/protect it from causing crashes, even if that means using a syntax that isn't always quite "Lua like".  Others have chosen to only provide limited access to limited features in their modules, ones that were easy/easier to make truly lua-like.  That's a design choice.

However, and perhaps I was more abrupt than I intended when I said this earlier (so apologies if you took it that way), I can code to expectations if someone has a strong opinion or there is the question of legacy code... I figured hs.settings.set has been around long enough, and since the original version which didn't use LuaSkin at all didn't suffer from there being no second argument to `set` (it was only when it was changed to use LuaSkin to simplify object conversion that it became a problem) supporting it if someone wanted it to isn't a problem.  It just went against my interpretation of the intention of the module, so I asked... and then made a comment about :-)

**Comment by lowne** [comment]
(Apologies for the frankly excessive length of this post 😞)

The divergence boils down to this (warning: blatant abuse of poetic license ahead): I *abhor* leaky abstractions, whereas you'd want *no abstractions* at all. 

I want many small puddles - in a hellish landscape full of sharp rocks and monstrous creatures - of thick, disgusting, toxic slime, so that I can't look into the multibranched abyss beneath even if I wanted to; whereas you want a very large, shallow and crystalline pool. As I'm busy running from the monsters I want to jump from puddle to puddle, dip in briefly as small a piece of my toe as possible (the slime is disgusting!), and get the job done quickly without even knowing where all the magic came from. You want to be able to wade freely and leisurely within the pool, umbrella cocktail in hand, not a worry in the world. 

Your pool is ObjC/OSX Cocoa: refreshingly cool, comfortable, a pleasure to be in. My nightmarish landscape is *all software ever made*, built with the sole purpose of making me suffer as much as possible.

To me, HS itself is clearly an *abstraction* over OSX, and I'd rather have it not leak all over the place. To wit:
> [theStyle=]hs.settings.get("AppleInterfaceStyle")

This has nothing to do with the idiom: `theStyle=hs.settings.AppleInterfaceStyle`, done.

> That's nowhere in Hammerspoon or Lua... that's in NSGlobalDomain

If there are good reasons to care at all whence it came, then `hs.settings.global.AppleInterfaceStyle`. But as a general rule I don't *want* to care. In fact, I'll leapfrog querying more foreign plists and one-up you: `hs.alert(hs.settings.temperatureAtAyersRock)` comes from a remote server in the centre of Australia. "Surely that can't be synchronous!" I hear you say; oh, it can, hold on a sec while I wrap my script in a coroutine-powered dispatcher (basically a non-leaky collection of Async monads) that takes care of all asynchronous calls for me... and done! My script will never *need* to care if the setting came from a plist on my disk or from Mars, and everything in it *looks* synchronous. Victory!

> it's an integral fact that Hammerspoon is not and never will be just Lua. It's a hybrid of Lua and a bridge to some of OS X

The way I see it, HS might be secretly powered by tiny hamsters that get "downloaded" inside my MBP, and I (the user) wouldn't care, as long as it does what it says on the box, which is, I *get to script it in Lua*. If the box says I'll need to use a "Lua-like", OSX-specific DSL, then fine, I'll (grudgingly) learn the DSL. Either way, I don't *want* to know *anything* about OSX, I want to read *just* the HS docs.

> I prefer giving access to as much as possible [...] Others have chosen to only provide limited access to limited features

That's a false dichotomy. I've advocated several times for *fully exposing* (but not necessarily fully documenting) the *naked*, low-level, ugly-syntax OSX calls, and *against* wrapping things in ObjC; so that then things can get wrapped in Lua (which *excels* at precisely this task) rather than in ObjC (which is *terrible*, if only because it's compiled), with a thought out, simple, idiomatic and *non-leaky* API - while still *keeping* full access to the lower levels for power users/module writers. Granted, most of this comes from the predecessors, but newer modules are still done the same way (and in some cases even worse). 
For a simple example of this (but they're really all over the place), see the "excessive" wrapping in `hs.timer.internal` that finally forced me to rewrap everything *again* for `timer.delayed`:

```lua
module.delayed = {
  new=function(delay,fn)
    local tmr=module.new(DISTANT_FUTURE,fn):start()
    return {
      start=function(self,dl) tmr:setNextTrigger(dl or delay) return self end,
      stop=function(self) tmr:setNextTrigger(DISTANT_FUTURE) return self end,
      nextTrigger=function() local nt=tmr:nextTrigger() return (nt>0 and nt<=delay) and nt or nil end,
      running=function(self) return self:nextTrigger() and true or false end,
      setDelay=function(self,dl) if self:running() then delay=dl self:start() end delay=dl return self end,
    }
  end
}
```

-----

This last point ties in nicely with what I believe is the crux of the matter:
> once it's decided the user API is meant to be used in ("pure") Lua

vs 
> anything whose ultimate destination is *not* the Lua interpreter, better not be vague

I unilaterally *define* the user as the person whose *sole* domain is the Lua interpreter, through a file called `~/.hammerspoon/init.lua`, so, as the saying goes, I'm right by definition there :)

I define the *module writer* as the person who writes *reusable* - as in, transferable to a different person with a different machine and a completely different `init.lua`, and still be useful - pieces of code, i.e. modules - in Lua, ObjC, or a combination.

(You'll notice there's no such thing as an *ObjC user*: "if you see a .so (or .dylib or whatever they're called) inside ~/.hammerspoon that can't be meaningfully transferred to another user, they blew it", as the quote goes.)

So we're left with the question: what's the domain of the *module writer* with regards to other modules? The de-facto answer is "everything goes in both Lua and ObjC"; therefore we *inevitably* end up with these "hybrid" APIs and the give-or-take struggle that you mention. But in my advocacy mentioned above, I dreamed of a different answer.

A clear way to express it is this simple rule: *from ObjC, you never call outside of your file, and you never call Lua*. This *forces* the ObjC side of modules to be nothing more than dumb lists of system calls (LuaSkin, which would mediate all callbacks, and internal pure utility functions are allowed of course); all the logic, bookkeeping etc. belong in Lua. What would happen if things were done this way?

1. it'd be *trivially* easy to create multi-layered APIs
* much easier to end up with a top API layer (the *user* layer) that is simple, idiomatic and leak-free *Lua*, as advertised on the box
* easier to improve and build upon existing functionality in useful ways by harnessing the middle and low layers, without having to *fight* against the "hybrid" APIs
* easier and quicker to open up entirely new abilities by writing thin(ner) ObjC bridges to OSX, and then organising them as per point above
* simple, obvious solutions to thorny issues such as the venerable #477 
* less userdata, less memory management, less crashes to debug
* simplified LuaSkin and surroundings
* drastically reduced need for recompilation during development (or really, for opening Xcode at all)
* a likely decrease in performance, which if detectable at all and deemed troublesome could be *more* than compensated once and for all by switching to LuaJIT+ffi (no easy task, admittedly)
* every single contributor current and future, with the exception of yours truly, would be very sad, because Lua is crazy and ObjC is awesome :)

I'm aware that point 10 vastly overpowers everything else, and I'm ok with it. I merely wanted to show that "otro mundo es (theoretically) posible", and many of your objections are *factually* true, but not *necessarily* true.

-----
>perhaps I was more abrupt than I intended

Not at all! As I said, I rather enjoy these apparently pointless discussions, and even hope that they "degenerate" into ever more polarised opposite stances, because they allow me to "prod" my "opponent"'s point of view from a lot of different sides, which, together with my "playing devil's advocate for the enemy" (or in other words, greatly exaggerating my own stance, as evidenced by the Australian server example), can unearth mistakes or fallacies in my own "beliefs" (that are otherwise in my blind spot, so to speak) and bring me to reconsider/adjust them. As for the original `settings.set` issue, I think it wouldn't be unfair to say that we're both satisfied with the proposed solution, if only because it maintains backward compatibility :)

**Comment by asmagill** [comment]
I laughed out loud at the implication that coroutines can mean asynchronous... at best coroutines can mimic async/non-breaking/threaded if either the process can be broken down into discrete separable steps, or you don't mind polling every so often to see if something else which *can* easily be asynchronous (in this case Obj-c/OS X) is ready and waiting until lua gets around to checking it.  Having coded/worked in languages designed to be multi-threaded before, coroutines seem limited to me and I usually avoid them... though I will admit that perhaps the right example could change my mind, but so far the ones I've seen seem tediously complex compared to how they might be done in another language, which does lead to a difference:  I say, use the language that makes the most sense and not make lua do what it really isn't meant to do.  But then I am and always will be a programmer and only "play a user on TV", as they say (though I do think I do a better job of it than some I've worked with before).

Polling is anathema to the original intention of Hydra/Mjolnir/Hammerspoon.  Not saying it should or shouldn't be, just that it was a design decision from the beginning.

And, I add so many callbacks for seemingly random things that are in the obj-c classes (well, seemingly random to me, until I actually realize I want to use them... but that usually happens well after I'm done writing the module and starting to use it and find that the possibilities are greater than what prompted me to write the module in the first place) specifically so most of the logic can be done in Lua, not Obj-C.

As for lua-jit... not compatible with Lua 5.3, and no intention on the maintainers part to make it so.  Not sure if there is a fork, though.  At any rate, losing the bit operators and the utf8 library (doable, but annoying), losing 64bit integers (less annoying except for a couple of specific cases that might not even be in core yet); undoing all of the integer vs. float changes to the modules we did when we changed from 5.2 to 5.3... I'll be one of the first in line to lynch whoever suggests it! ;-)

the rest... I'll have to re-read when I have time and let it sink in.

**Comment by heptal** [comment]
I'd just like to add that I like the approach taken so far. Lua's killer feature is that it easily adapts to and extends the environment it is in. I feel comfortable saying that 'idiomatic Lua' is entirely context dependent: of course Hammerspoon is an 'OS X-specific DSL' and doesn't imply otherwise. From what I understand, a primary reason Hammerspoon was forked from mjolnir was to provide 'out-of-the-box' high level functionality precisely so that users wouldn't have to locate, maintain, or implement complex abstraction layers to accomplish OS X-specific tasks. It was made to be useful and powerful and issues with the expressiveness of the API are usually resolved. 

In my opinion, it is faster, easier, and more concise to manipulate the OS X side of things with Obj-C (or Swift) in a lot of cases. Would you really want to undertake the laborious process of deconstructing a (usually, but not always, well-designed) OS X/Obj-C API into tiny Lua accessors and rebuilding it in Lua without the help of the typing system or Apple framework idioms that people know and like? It's amazing what your compiler can do for you. I would much rather do a lot of 'high-level' image manipulation operations in Obj-C, calling on other modules as needed, and exposing the most useful parameters to Lua. More can always be added later. I feel like this project does a great job keeping the parts best-expressible in either Obj-C or Lua in their respective domains, and is a big reason people are attracted to this project.

**Comment by cmsj** [comment]
I have no idea anymore what is happening on this issue, but it doesn't feel very productive at this point :)

Please reopen it if there are actual actions to be taken, otherwise I suggest that philosophy is better discussed on the mailing list.

**Comment by asmagill** [comment]
(just seeing if you're still alive, @cmsj 😝)

**Comment by lowne** [comment]
An example of action that *could* be taken in my ideal world:

Since... 
1. RGB has been *the* way to indicate colours on computers since before the "invention" of 24-bit capable screens
2. HSV/HSL, CMYK and what-have-you are used/useful in specific domains, and at some layer down the abstraction cake (e.g. OpenGL) conversion to RGB is performed anyways
3. I don't want to have to type `{red=R,blue=B,green=G,alpha=A}` a zillion times when obviously `{R,G,B,A}` can and should do 
4. `hs.grid`, `hs.expose` and soon `hs.window.switcher` and `hs.window.highlight` all expose a `ui` field with tons of user-customisable colours
5. I also don't want *users* to have to type the longform as per point 3 above

... I've had a `getColor(t)` function in each of the aforementioned modules. I had found and fixed an unrelated bug (which itself only exists because I have to jump through hoops to begin with, in order to enable point 5 above) in one of the modules. Just now I noticed the same problem on another module, and while I clearly remembered that I had fixed it already (and recently), I had to re-diagnose it, and then remember to perform identical changes *in all 4 modules*. 

Sounds like an obvious candidate for DRY, yes? Well, too bad, because apparently Apple decided that NSColor should be completely colour-space agnostic, and HS is so tightly coupled to NSColor that, afaict, there's no (sane) way to drive the slightest wedge between the two in order to introduce an however thin layer in-between.

You'll probably need to squint a bit, but you should see that this is the same kind of thing that happens with `hs.timer` (to resurrect and clarify on that example), where a userdata is created with the sole purpose of feeding a "context" to the actual timer's constructor (and Apple thought of that because it's convenient to have it *in ObjC*) and then juggled around in internal.m - as opposed to just passing the timer itself back (from the constructor) and forth as a naked pointer, and building the object itself and the API (which has to be "redone" over "Apple's one true way" *either way* since we *do* need to expose it) in Lua - which (to me at least) seems a lot more straightforward, painlessly extensible etc. (see all my posts above). (As for type safety and compiler help, I'm the world's biggest fan of it, but even if it helped in these instances (and it doesn't *at all*), I'll take fast iteration/immediate feedback/live coding over it any day... but I'll cut it short)

So this issue will remain closed, but I see no harm in keeping it as my rant box :p (fwiw this is the first I hear of the mailing list, and it's not clear immediately clear what its purpose is; I'd say that "philosophy" about *development* should be discussed where development actually takes place, i.e. here)

**Comment by lowne** [comment]
Contrary to my previous post, it should be possible to insert implementation for {R,G,B,A} right [here](https://github.com/Hammerspoon/hammerspoon/blob/master/extensions/drawing/color/internal.m#L185)
